### PR TITLE
Add support for loading auth clients from local.properties

### LINKED ISSUE #636: Support switching of Oauth clients
Each environment i.e staging, preview or production may have different Oauth clients.
We need the ability to switch the Oauth clients used for authentication so that a single app can authenticate on different environments.

- [ ] To allow easy configuration of multiple clients in this standard being being proposed when defining oauth2 clients in local.properties

```
oauth.client.{env}.id="client_id"
oauth.client.{{env}}.secret="client_secret"
```

- [ ] The env will be pointing to the environment that the client is for.

- [ ] Add functionality for loading all properties starting with `oauth.client` to a map/bean/properties object in core

- [ ] Implement a Map that stores the mapping of the environment used on local.properties to host names

- [ ] Add listener on the settings activity so that when the base URL is updated the correct, the URL is used to fetch the environment. If the lookup fails then the a dialog should be shown informing the user that the app is not configured so access that URL.

- [ ] When making requests the app should be able to resolve clients details depending on host being used

### PR DESCRIPTION
**Pull request by faidvd**
Resolves #636

### CODE REVIEW

File: opensrp-app/src/main/java/org/smartregister/domain/Environment.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+
+public class Environment {
+    @SerializedName("env")
+    @Expose
+    protected String env;
+
+    @SerializedName("id")
```
Comment (ekigamba) [inline opensrp-app/src/main/java/org/smartregister/domain/Environment.java:11]: `client.id` would be better

File: opensrp-app/src/main/java/org/smartregister/domain/Environment.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    @Expose
+    protected String id;
+
+    @SerializedName("url")
+    @Expose
+    protected String url;
+
+    @SerializedName("secret")
```
Comment (ekigamba) [inline opensrp-app/src/main/java/org/smartregister/domain/Environment.java:19]: `client.secret` would be better

**Review by ekigamba — CHANGES_REQUESTED** [review]

**Comment by ekigamba** [comment]
Looks good to me

**Comment by Rkareko** [comment]
@faidvd please bump the version

**Comment by Rkareko** [comment]
@faidvd can we have some javadocs for the public methods

**Review by ndegwamartin — COMMENTED** [review]
Can we add some java docs for the new classes and methods?

File: opensrp-app/src/main/java/org/smartregister/view/activity/SettingsActivity.java [hunk 3]
```diff
@@ -92,7 +92,7 @@ public boolean onPreferenceClick(Preference preference) {
     public void onClick(View view) {
 
         String newValue = baseUrlEditTextPreference.getEditText().getText().toString();
-        if (newValue != null && UrlUtil.isValidUrl(newValue)) {
+        if (newValue != null && UrlUtil.isValidUrl(newValue) && UrlUtil.isValidEnvironment(newValue)) {
```
Comment (ndegwamartin) [inline opensrp-app/src/main/java/org/smartregister/view/activity/SettingsActivity.java:95]: Should the order here be changed?

File: gradle.properties [hunk 4]
```diff
@@ -1,4 +1,4 @@
-VERSION_NAME=4.2.18-SNAPSHOT
+VERSION_NAME=4.2.19-SNAPSHOT
```
Comment (Rkareko) [inline gradle.properties:1]: Let's bump this up to `4.3.0` since it's new backward compatible feature

**Review by Rkareko — APPROVED** [review]

**Review by ekigamba — APPROVED** [review]

**Pull request merged**